lgc/elfLinker/FetchShader.cpp

```diff
@@ -114,42 +136,104 @@ Module *FetchShader::generate() {
+// @param call : A call to the InstanceId shader input builtin.
+// @returns : The value of the InstanceId.  All new code will be place at the start of the function containing call.
+Value *FetchShader::getReplacementForInstanceIdBuiltIn(CallInst *call) const {
+  Function *callerFunction = call->getParent()->getParent();
+  return getVgprArgumentAsAnInt32(m_vsEntryRegInfo.instanceId, callerFunction);
+}
+
+// =====================================================================================================================
+// @param call : A call to the VertexId shader input builtin.
+// @returns : The value of the VertexId.  All new code will be place at the start of the function containing call.
+Value *FetchShader::getReplacementForVertexIdBuiltIn(CallInst *call) const {
+  Function *callerFunction = call->getParent()->getParent();
+  return getVgprArgumentAsAnInt32(m_vsEntryRegInfo.vertexId, callerFunction);
+}
+
+// =====================================================================================================================
+// @param vgpr : The vgpr number from which the value should come.
+// @param function : The function from which to get the argument.
+// @returns : The value of the argument in the function that corresponds to the given vgpr cast to a 32-bit integer.
+Value *FetchShader::getVgprArgumentAsAnInt32(unsigned int vgpr, Function *function) const {
+  BuilderBase builder(&*function->front().getFirstInsertionPt());
+  Value *vertexId = getVpgrArgument(vgpr, builder);
+  return builder.CreateBitCast(vertexId, builder.getInt32Ty());
+}
+
+// =====================================================================================================================
+// @param vgpr : The vgpr number from which the value should come.
+// @param builder : The builder to use if new instructions are needed.
+// @returns : The value of the argument in the function the that insertion point of the build that corresponds to the
+// given vgpr.
+Value *FetchShader::getVpgrArgument(unsigned int vgpr, BuilderBase &builder) const {
+  Function *function = builder.GetInsertPoint()->getFunction();
+  if (!mustFixLsVgprInput())
+    return function->getArg(m_vsEntryRegInfo.sgprCount + vgpr);
+
+  Type *int32Type = builder.getInt32Ty();
+  Value *mergeWaveInfo = function->getArg(LsHsSysValueMergedWaveInfo);
+  Value *eight = builder.getInt32(8);
+  std::vector<Value *> args = {mergeWaveInfo, eight, eight};
+  std::vector<Attribute::AttrKind> attribs = {Attribute::ReadNone};
+  Value *hsVertexCount = builder.CreateNamedCall("llvm.amdgcn.ubfe.i32", int32Type, args, attribs, "HsVertCount");
+  Value *isNullHs = builder.CreateICmp(llvm::CmpInst::ICMP_EQ, hsVertexCount, builder.getInt32(0), "IsNullHs");
+
+  Value *valueForNonNullHs = function->getArg(m_vsEntryRegInfo.sgprCount + vgpr);
+  Value *valueForNullHs = function->getArg(m_vsEntryRegInfo.sgprCount + vgpr - 2);
```

Comment:
Should explain the need for ```-2 ``.

Reply:
I added a comment further up that explains the bug and I replaced the 2 with a named constant.